File: examples/nodejs/MijAuth.js
/**
 * MijAuth - System Weryfikacji Dwuetapowej (2FA) oparty na plikach
 * Implementacja Node.js (JavaScript/TypeScript)
 * 
 * Wymaga Node.js 16+ (używa wbudowanego modułu crypto)
 */

const crypto = require('crypto');
const fs = require('fs');

class MijAuth {
    static ALGORITHM = 'aes-256-gcm';
    static KEY_LENGTH = 32; // 256 bits
    static IV_LENGTH = 12;  // 96 bits dla GCM
    static TAG_LENGTH = 16; // 128 bits
    static VERSION = 1;
    static DEFAULT_AUTH_FILE_TTL_SECONDS = 30 * 24 * 60 * 60;

    /**
     * Generuje nowy klucz AES-256 dla użytkownika
     * @returns {string} Klucz w formacie base64
     */
    static generateUserKey() {
        const key = crypto.randomBytes(this.KEY_LENGTH);
        return key.toString('base64');
    }

    /**
     * Generuje unikalny token dla użytkownika
     * @returns {string} Token w formacie hex
     */
    static generateToken() {
        return crypto.randomBytes(32).toString('hex');
    }

    /**
     * Tworzy zaszyfrowany plik autoryzacyjny .mijauth
     * 
     * @param {string} userId - Identyfikator użytkownika
     * @param {string} userKeyBase64 - Klucz użytkownika w base64
     * @param {string|null} deviceHash - Opcjonalny hash urządzenia
     * @returns {{fileContent: string, token: string}}
     */
    static createAuthFile(userId, userKeyBase64, deviceHash = null, deviceHashV2 = null) {
        const token = this.generateToken();
        
        const payload = {
            user_id: userId,
            token: token,
            created_at: new Date().toISOString(),
            device_hash: deviceHash,
            device_hash_v2: deviceHashV2,
            version: this.VERSION
        };

        const jsonPayload = JSON.stringify(payload);
        const encryptedContent = this.encrypt(jsonPayload, userKeyBase64);

        return {
            fileContent: encryptedContent,
            token: token
        };
    }

    /**
     * Weryfikuje plik autoryzacyjny i zwraca dane użytkownika
     * 
     * @param {string} fileContent - Zawartość pliku .mijauth
     * @param {string} userKeyBase64 - Klucz użytkownika w base64
     * @returns {object|null} Dane użytkownika lub null
     */
    static verifyAuthFile(fileContent, userKeyBase64, maxAgeSeconds = this.DEFAULT_AUTH_FILE_TTL_SECONDS) {
        try {
            const decrypted = this.decrypt(fileContent, userKeyBase64);
            
            if (decrypted === null) {
                return null;
            }

            const payload = JSON.parse(decrypted);
            
            // Walidacja struktury
            if (!payload.user_id || !payload.token || !payload.version) {
                return null;
            }

            if (!this.isAuthFileWithinTtl(payload, maxAgeSeconds)) {
                return null;
            }

            return payload;
        } catch (e) {
            return null;
        }
    }

    /**
     * Weryfikuje plik i sprawdza czy token zgadza się z przechowywanym
     * 
     * @param {string} fileContent - Zawartość pliku .mijauth
     * @param {string} userKeyBase64 - Klucz użytkownika w base64
     * @param {string} expectedToken - Oczekiwany token z bazy danych
     * @param {string} expectedUserId - Oczekiwane ID użytkownika
     * @returns {boolean}
     */
    static verifyAuthFileWithToken(
        fileContent,
        userKeyBase64,
        expectedToken,
        expectedUserId,
        maxAgeSeconds = this.DEFAULT_AUTH_FILE_TTL_SECONDS
    ) {
        const payload = this.verifyAuthFile(fileContent, userKeyBase64, maxAgeSeconds);
        
        if (payload === null) {
            return false;
        }

        // Constant-time comparison
        const tokenMatch = crypto.timingSafeEqual(
            Buffer.from(expectedToken),
            Buffer.from(payload.token)
        );
        const userIdMatch = crypto.timingSafeEqual(
            Buffer.from(expectedUserId),
            Buffer.from(payload.user_id)
        );

        return tokenMatch && userIdMatch;
    }

    /**
     * Weryfikuje plik, token i hash urządzenia (v1/v2)
     *
     * @param {string} fileContent
     * @param {string} userKeyBase64
     * @param {string} expectedToken
     * @param {string} expectedUserId
     * @param {string|null} expectedDeviceHash
     * @param {string|null} expectedDeviceHashV2
     * @returns {boolean}
     */
    static verifyAuthFileWithTokenAndDevice(
        fileContent,
        userKeyBase64,
        expectedToken,
        expectedUserId,
        expectedDeviceHash = null,
        expectedDeviceHashV2 = null
    ) {
        const payload = this.verifyAuthFile(fileContent, userKeyBase64);

        if (payload === null) {
            return false;
        }

        const tokenMatch = crypto.timingSafeEqual(
            Buffer.from(expectedToken),
            Buffer.from(payload.token)
        );
        const userIdMatch = crypto.timingSafeEqual(
            Buffer.from(expectedUserId),
            Buffer.from(payload.user_id)
        );

        if (!tokenMatch || !userIdMatch) {
            return false;
        }

        if (expectedDeviceHash !== null) {
            if (!payload.device_hash) {
                return false;
            }
            const deviceMatch = crypto.timingSafeEqual(
                Buffer.from(expectedDeviceHash),
                Buffer.from(payload.device_hash)
            );
            if (!deviceMatch) {
                return false;
            }
        }

        if (expectedDeviceHashV2 !== null) {
            if (!payload.device_hash_v2) {
                return false;
            }
            const deviceMatchV2 = crypto.timingSafeEqual(
                Buffer.from(expectedDeviceHashV2),
                Buffer.from(payload.device_hash_v2)
            );
            if (!deviceMatchV2) {
                return false;
            }
        }

        return true;
    }

    /**
     * Regeneruje plik autoryzacyjny (nowy token)
     * 
     * @param {string} userId - Identyfikator użytkownika
     * @param {string} userKeyBase64 - Klucz użytkownika w base64
     * @param {string|null} deviceHash - Opcjonalny hash urządzenia
     * @returns {{fileContent: string, token: string}}
     */
    static regenerateAuthFile(userId, userKeyBase64, deviceHash = null, deviceHashV2 = null) {
        return this.createAuthFile(userId, userKeyBase64, deviceHash, deviceHashV2);
    }

    /**
     * Szyfruje dane przy użyciu AES-256-GCM
     * 
     * @param {string} plaintext - Dane do zaszyfrowania
     * @param {string} keyBase64 - Klucz w formacie base64
     * @returns {string} Zaszyfrowane dane w formacie base64
     */
    static encrypt(plaintext, keyBase64) {
        const key = Buffer.from(keyBase64, 'base64');
        const iv = crypto.randomBytes(this.IV_LENGTH);

        const cipher = crypto.createCipheriv(this.ALGORITHM, key, iv);
        
        let ciphertext = cipher.update(plaintext, 'utf8');
        ciphertext = Buffer.concat([ciphertext, cipher.final()]);
        
        const tag = cipher.getAuthTag();

        // Format: IV (12 bytes) + Tag (16 bytes) + Ciphertext
        const combined = Buffer.concat([iv, tag, ciphertext]);
        
        return combined.toString('base64');
    }

    /**
     * Odszyfrowuje dane przy użyciu AES-256-GCM
     * 
     * @param {string} encryptedBase64 - Zaszyfrowane dane w formacie base64
     * @param {string} keyBase64 - Klucz w formacie base64
     * @returns {string|null} Odszyfrowane dane lub null przy błędzie
     */
    static decrypt(encryptedBase64, keyBase64) {
        try {
            const key = Buffer.from(keyBase64, 'base64');
            const combined = Buffer.from(encryptedBase64, 'base64');

            if (combined.length < this.IV_LENGTH + this.TAG_LENGTH) {
                return null;
            }

            const iv = combined.subarray(0, this.IV_LENGTH);
            const tag = combined.subarray(this.IV_LENGTH, this.IV_LENGTH + this.TAG_LENGTH);
            const ciphertext = combined.subarray(this.IV_LENGTH + this.TAG_LENGTH);

            const decipher = crypto.createDecipheriv(this.ALGORITHM, key, iv);
            decipher.setAuthTag(tag);

            let plaintext = decipher.update(ciphertext);
            plaintext = Buffer.concat([plaintext, decipher.final()]);

            return plaintext.toString('utf8');
        } catch (e) {
            return null;
        }
    }

    /**
     * Generuje hash urządzenia na podstawie headers (dla Express.js)
     * 
     * @param {object} headers - Obiekt nagłówków HTTP
     * @returns {string} Hash urządzenia
     */
    static generateDeviceHash(headers) {
        const data = {
            user_agent: headers['user-agent'] || '',
            accept_language: headers['accept-language'] || '',
            device_id: headers['x-mijauth-device-id'] || ''
        };

        return crypto.createHash('sha256')
            .update(JSON.stringify(data))
            .digest('hex');
    }

    /**
     * Generuje hash urządzenia v2 z rozszerzonego kontekstu
     *
     * @param {object} context
     * @returns {string}
     */
    static generateDeviceHashV2(context = {}) {
        const normalized = this._normalizeFingerprintContext(context);
        return crypto.createHash('sha256')
            .update(JSON.stringify(normalized))
            .digest('hex');
    }

    static _normalizeFingerprintContext(context) {
        const normalized = {};
        Object.keys(context)
            .sort()
            .forEach((key) => {
                const value = context[key];
                if (value === null || value === undefined) {
                    return;
                }
                if (typeof value === 'object' && !Array.isArray(value)) {
                    normalized[key] = this._normalizeFingerprintContext(value);
                } else {
                    normalized[key] = value;
                }
            });
        return normalized;
    }

    static generateTotpSecret(length = 32) {
        if (length < 16) {
            throw new Error('TOTP secret must have at least 16 Base32 chars');
        }
        const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
        const bytes = crypto.randomBytes(length);
        let secret = '';
        for (let i = 0; i < length; i++) {
            secret += alphabet[bytes[i] % 32];
        }
        return secret;
    }

    static getTotpProvisioningUri(accountName, issuer, secret, digits = 6, period = 30) {
        const label = encodeURIComponent(`${issuer}:${accountName}`);
        const query = new URLSearchParams({
            secret: secret.toUpperCase(),
            issuer,
            algorithm: 'SHA1',
            digits: String(digits),
            period: String(period)
        }).toString();
        return `otpauth://totp/${label}?${query}`;
    }

    static generateTotpCode(secret, timestamp = Date.now(), period = 30, digits = 6) {
        const counter = Math.floor(timestamp / 1000 / period);
        return this.generateHotpCode(secret, counter, digits);
    }

    static verifyTotp(secret, code, discrepancy = 1, timestamp = Date.now(), period = 30, digits = 6) {
        const normalizedCode = String(code).replace(/\s+/g, '');
        if (!new RegExp(`^\\d{${digits}}$`).test(normalizedCode)) {
            return false;
        }

        const counter = Math.floor(timestamp / 1000 / period);
        for (let offset = -discrepancy; offset <= discrepancy; offset++) {
            const candidate = this.generateHotpCode(secret, counter + offset, digits);
            if (crypto.timingSafeEqual(Buffer.from(candidate), Buffer.from(normalizedCode))) {
                return true;
            }
        }
        return false;
    }

    static generateHotpCode(secret, counter, digits = 6) {
        if (counter < 0) {
            return '0'.repeat(digits);
        }
        const secretBuffer = this.base32Decode(secret);
        if (!secretBuffer || secretBuffer.length === 0) {
            throw new Error('Invalid TOTP secret');
        }

        const counterBuffer = Buffer.alloc(8);
        counterBuffer.writeUInt32BE(Math.floor(counter / 0x100000000), 0);
        counterBuffer.writeUInt32BE(counter >>> 0, 4);

        const hash = crypto.createHmac('sha1', secretBuffer).update(counterBuffer).digest();
        const offset = hash[hash.length - 1] & 0x0f;
        const binary =
            ((hash[offset] & 0x7f) << 24) |
            ((hash[offset + 1] & 0xff) << 16) |
            ((hash[offset + 2] & 0xff) << 8) |
            (hash[offset + 3] & 0xff);

        return String(binary % (10 ** digits)).padStart(digits, '0');
    }

    static base32Decode(secret) {
        const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
        const map = new Map(alphabet.split('').map((char, idx) => [char, idx]));
        const normalized = String(secret).toUpperCase().replace(/=|\s+/g, '');
        let bits = '';

        for (const char of normalized) {
            if (!map.has(char)) {
                return Buffer.alloc(0);
            }
            bits += map.get(char).toString(2).padStart(5, '0');
        }

        const bytes = [];
        for (let i = 0; i + 8 <= bits.length; i += 8) {
            bytes.push(parseInt(bits.slice(i, i + 8), 2));
        }

        return Buffer.from(bytes);
    }

    static isAuthFileWithinTtl(payload, maxAgeSeconds) {
        if (maxAgeSeconds === null || maxAgeSeconds === undefined) {
            return true;
        }

        const createdAt = Date.parse(payload.created_at || '');
        if (Number.isNaN(createdAt)) {
            return false;
        }

        const now = Date.now();
        if (createdAt > now) {
            return false;
        }

        return (now - createdAt) <= maxAgeSeconds * 1000;
    }
}

/**
 * Symulacja bazy danych użytkowników
 */
class UserDatabase {
    constructor(storageFile = 'users.json') {
        this.storageFile = storageFile;
        this.users = {};
        this.load();
    }

    load() {
        try {
            if (fs.existsSync(this.storageFile)) {
                const data = fs.readFileSync(this.storageFile, 'utf8');
                this.users = JSON.parse(data);
            }
        } catch (e) {
            this.users = {};
        }
    }

    save() {
        fs.writeFileSync(this.storageFile, JSON.stringify(this.users, null, 2));
    }

    async createUser(userId, email, password) {
        const userKey = MijAuth.generateUserKey();
        const authResult = MijAuth.createAuthFile(userId, userKey);
        const totpSecret = MijAuth.generateTotpSecret();

        // Hashowanie hasła (prosty przykład - w produkcji użyj bcrypt/argon2)
        const passwordHash = crypto.createHash('sha256')
            .update(password + 'salt_' + userId)
            .digest('hex');

        this.users[userId] = {
            id: userId,
            email: email,
            password_hash: passwordHash,
            encryption_key: userKey,
            auth_token: authResult.token,
            totp_secret: totpSecret,
            created_at: new Date().toISOString()
        };

        this.save();

        return {
            user: this.users[userId],
            authFile: authResult.fileContent
        };
    }

    getUser(userId) {
        return this.users[userId] || null;
    }

    getUserByEmail(email) {
        return Object.values(this.users).find(u => u.email === email) || null;
    }

    verifyPassword(user, password) {
        const hash = crypto.createHash('sha256')
            .update(password + 'salt_' + user.id)
            .digest('hex');
        return crypto.timingSafeEqual(Buffer.from(hash), Buffer.from(user.password_hash));
    }

    updateAuthToken(userId, newToken) {
        if (this.users[userId]) {
            this.users[userId].auth_token = newToken;
            this.save();
        }
    }
}

module.exports = { MijAuth, UserDatabase };

File: examples/dotnet/MijAuth.cs
/**
 * MijAuth - System Weryfikacji Dwuetapowej (2FA) oparty na plikach
 * Implementacja .NET (C#)
 * 
 * Wymaga .NET 6.0+
 */

using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;

namespace MijAuth
{
    /// <summary>
    /// Główna klasa systemu MijAuth do weryfikacji dwuetapowej
    /// </summary>
    public static class MijAuthService
    {
        private const int KeyLength = 32;  // 256 bits
        private const int IvLength = 12;   // 96 bits dla GCM
        private const int TagLength = 16;  // 128 bits
        private const int Version = 1;
        private const int DefaultAuthFileTtlSeconds = 30 * 24 * 60 * 60;

        /// <summary>
        /// Generuje nowy klucz AES-256 dla użytkownika
        /// </summary>
        /// <returns>Klucz w formacie base64</returns>
        public static string GenerateUserKey()
        {
            var key = new byte[KeyLength];
            RandomNumberGenerator.Fill(key);
            return Convert.ToBase64String(key);
        }

        /// <summary>
        /// Generuje unikalny token dla użytkownika
        /// </summary>
        /// <returns>Token w formacie hex</returns>
        public static string GenerateToken()
        {
            var token = new byte[32];
            RandomNumberGenerator.Fill(token);
            return Convert.ToHexString(token).ToLowerInvariant();
        }

        /// <summary>
        /// Tworzy zaszyfrowany plik autoryzacyjny .mijauth
        /// </summary>
        /// <param name="userId">Identyfikator użytkownika</param>
        /// <param name="userKeyBase64">Klucz użytkownika w base64</param>
        /// <param name="deviceHash">Opcjonalny hash urządzenia</param>
        /// <returns>Tuple z zawartością pliku i tokenem</returns>
        public static (string FileContent, string Token) CreateAuthFile(
            string userId,
            string userKeyBase64,
            string? deviceHash = null,
            string? deviceHashV2 = null)
        {
            var token = GenerateToken();

            var payload = new AuthPayload
            {
                UserId = userId,
                Token = token,
                CreatedAt = DateTime.UtcNow.ToString("o"),
                DeviceHash = deviceHash,
                DeviceHashV2 = deviceHashV2,
                Version = Version
            };

            var jsonPayload = JsonSerializer.Serialize(payload);
            var encryptedContent = Encrypt(jsonPayload, userKeyBase64);

            return (encryptedContent, token);
        }

        /// <summary>
        /// Weryfikuje plik autoryzacyjny i zwraca dane użytkownika
        /// </summary>
        /// <param name="fileContent">Zawartość pliku .mijauth</param>
        /// <param name="userKeyBase64">Klucz użytkownika w base64</param>
        /// <returns>Dane użytkownika lub null</returns>
        public static AuthPayload? VerifyAuthFile(string fileContent, string userKeyBase64, int? maxAgeSeconds = DefaultAuthFileTtlSeconds)
        {
            try
            {
                var decrypted = Decrypt(fileContent, userKeyBase64);

                if (decrypted == null)
                    return null;

                var payload = JsonSerializer.Deserialize<AuthPayload>(decrypted);

                // Walidacja struktury
                if (payload == null || 
                    string.IsNullOrEmpty(payload.UserId) || 
                    string.IsNullOrEmpty(payload.Token))
                {
                    return null;
                }

                if (maxAgeSeconds.HasValue)
                {
                    if (!DateTimeOffset.TryParse(payload.CreatedAt, out var createdAt))
                        return null;

                    var now = DateTimeOffset.UtcNow;
                    if (createdAt > now)
                        return null;

                    if ((now - createdAt).TotalSeconds > maxAgeSeconds.Value)
                        return null;
                }

                return payload;
            }
            catch
            {
                return null;
            }
        }

        /// <summary>
        /// Weryfikuje plik i sprawdza czy token zgadza się z przechowywanym
        /// </summary>
        /// <param name="fileContent">Zawartość pliku .mijauth</param>
        /// <param name="userKeyBase64">Klucz użytkownika w base64</param>
        /// <param name="expectedToken">Oczekiwany token z bazy danych</param>
        /// <param name="expectedUserId">Oczekiwane ID użytkownika</param>
        /// <returns>True jeśli weryfikacja pomyślna</returns>
        public static bool VerifyAuthFileWithToken(
            string fileContent,
            string userKeyBase64,
            string expectedToken,
            string expectedUserId,
            int? maxAgeSeconds = DefaultAuthFileTtlSeconds)
        {
            var payload = VerifyAuthFile(fileContent, userKeyBase64, maxAgeSeconds);

            if (payload == null)
                return false;

            // Constant-time comparison
            return CryptographicOperations.FixedTimeEquals(
                       Encoding.UTF8.GetBytes(expectedToken),
                       Encoding.UTF8.GetBytes(payload.Token)) &&
                   CryptographicOperations.FixedTimeEquals(
                       Encoding.UTF8.GetBytes(expectedUserId),
                       Encoding.UTF8.GetBytes(payload.UserId));
        }

        /// <summary>
        /// Generuje sekret TOTP Base32
        /// </summary>
        public static string GenerateTotpSecret(int length = 32)
        {
            if (length < 16)
                throw new ArgumentException("Sekret TOTP musi mieć co najmniej 16 znaków", nameof(length));

            const string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
            var bytes = RandomNumberGenerator.GetBytes(length);
            var chars = new char[length];

            for (var i = 0; i < length; i++)
                chars[i] = alphabet[bytes[i] % alphabet.Length];

            return new string(chars);
        }

        /// <summary>
        /// Tworzy URI otpauth do sparowania aplikacji
        /// </summary>
        public static string GetTotpProvisioningUri(string accountName, string issuer, string secret, int digits = 6, int period = 30)
        {
            var label = Uri.EscapeDataString($"{issuer}:{accountName}");
            var query = $"secret={Uri.EscapeDataString(secret.ToUpperInvariant())}" +
                        $"&issuer={Uri.EscapeDataString(issuer)}" +
                        "&algorithm=SHA1" +
                        $"&digits={digits}" +
                        $"&period={period}";

            return $"otpauth://totp/{label}?{query}";
        }

        /// <summary>
        /// Generuje kod TOTP
        /// </summary>
        public static string GenerateTotpCode(string secret, DateTimeOffset? timestamp = null, int period = 30, int digits = 6)
        {
            var ts = timestamp ?? DateTimeOffset.UtcNow;
            var counter = ts.ToUnixTimeSeconds() / period;
            return GenerateHotpCode(secret, counter, digits);
        }

        /// <summary>
        /// Weryfikuje kod TOTP
        /// </summary>
        public static bool VerifyTotp(string secret, string code, int discrepancy = 1, DateTimeOffset? timestamp = null, int period = 30, int digits = 6)
        {
            var normalizedCode = code.Replace(" ", "", StringComparison.Ordinal);
            if (normalizedCode.Length != digits || !normalizedCode.All(char.IsDigit))
                return false;

            var ts = timestamp ?? DateTimeOffset.UtcNow;
            var counter = ts.ToUnixTimeSeconds() / period;

            for (var offset = -discrepancy; offset <= discrepancy; offset++)
            {
                var candidate = GenerateHotpCode(secret, counter + offset, digits);
                if (CryptographicOperations.FixedTimeEquals(
                        Encoding.UTF8.GetBytes(candidate),
                        Encoding.UTF8.GetBytes(normalizedCode)))
                {
                    return true;
                }
            }

            return false;
        }

        /// <summary>
        /// Weryfikuje plik, token i hash urządzenia (v1/v2)
        /// </summary>
        public static bool VerifyAuthFileWithTokenAndDevice(
            string fileContent,
            string userKeyBase64,
            string expectedToken,
            string expectedUserId,
            string? expectedDeviceHash = null,
            string? expectedDeviceHashV2 = null)
        {
            var payload = VerifyAuthFile(fileContent, userKeyBase64);

            if (payload == null)
                return false;

            var tokenMatch = CryptographicOperations.FixedTimeEquals(
                Encoding.UTF8.GetBytes(expectedToken),
                Encoding.UTF8.GetBytes(payload.Token));

            var userMatch = CryptographicOperations.FixedTimeEquals(
                Encoding.UTF8.GetBytes(expectedUserId),
                Encoding.UTF8.GetBytes(payload.UserId));

            if (!tokenMatch || !userMatch)
                return false;

            if (expectedDeviceHash != null)
            {
                if (string.IsNullOrEmpty(payload.DeviceHash))
                    return false;

                if (!CryptographicOperations.FixedTimeEquals(
                        Encoding.UTF8.GetBytes(expectedDeviceHash),
                        Encoding.UTF8.GetBytes(payload.DeviceHash)))
                    return false;
            }

            if (expectedDeviceHashV2 != null)
            {
                if (string.IsNullOrEmpty(payload.DeviceHashV2))
                    return false;

                if (!CryptographicOperations.FixedTimeEquals(
                        Encoding.UTF8.GetBytes(expectedDeviceHashV2),
                        Encoding.UTF8.GetBytes(payload.DeviceHashV2)))
                    return false;
            }

            return true;
        }

        /// <summary>
        /// Regeneruje plik autoryzacyjny (nowy token)
        /// </summary>
        /// <param name="userId">Identyfikator użytkownika</param>
        /// <param name="userKeyBase64">Klucz użytkownika w base64</param>
        /// <param name="deviceHash">Opcjonalny hash urządzenia</param>
        /// <returns>Tuple z zawartością pliku i tokenem</returns>
        public static (string FileContent, string Token) RegenerateAuthFile(
            string userId,
            string userKeyBase64,
            string? deviceHash = null,
            string? deviceHashV2 = null)
        {
            return CreateAuthFile(userId, userKeyBase64, deviceHash, deviceHashV2);
        }

        /// <summary>
        /// Szyfruje dane przy użyciu AES-256-GCM
        /// </summary>
        private static string Encrypt(string plaintext, string keyBase64)
        {
            var key = Convert.FromBase64String(keyBase64);
            var iv = new byte[IvLength];
            RandomNumberGenerator.Fill(iv);

            var plaintextBytes = Encoding.UTF8.GetBytes(plaintext);
            var ciphertext = new byte[plaintextBytes.Length];
            var tag = new byte[TagLength];

            using var aes = new AesGcm(key, TagLength);
            aes.Encrypt(iv, plaintextBytes, ciphertext, tag);

            // Format: IV (12 bytes) + Tag (16 bytes) + Ciphertext
            var combined = new byte[iv.Length + tag.Length + ciphertext.Length];
            Buffer.BlockCopy(iv, 0, combined, 0, iv.Length);
            Buffer.BlockCopy(tag, 0, combined, iv.Length, tag.Length);
            Buffer.BlockCopy(ciphertext, 0, combined, iv.Length + tag.Length, ciphertext.Length);

            return Convert.ToBase64String(combined);
        }

        /// <summary>
        /// Odszyfrowuje dane przy użyciu AES-256-GCM
        /// </summary>
        private static string? Decrypt(string encryptedBase64, string keyBase64)
        {
            try
            {
                var key = Convert.FromBase64String(keyBase64);
                var combined = Convert.FromBase64String(encryptedBase64);

                if (combined.Length < IvLength + TagLength)
                    return null;

                var iv = new byte[IvLength];
                var tag = new byte[TagLength];
                var ciphertext = new byte[combined.Length - IvLength - TagLength];

                Buffer.BlockCopy(combined, 0, iv, 0, IvLength);
                Buffer.BlockCopy(combined, IvLength, tag, 0, TagLength);
                Buffer.BlockCopy(combined, IvLength + TagLength, ciphertext, 0, ciphertext.Length);

                var plaintext = new byte[ciphertext.Length];

                using var aes = new AesGcm(key, TagLength);
                aes.Decrypt(iv, ciphertext, tag, plaintext);

                return Encoding.UTF8.GetString(plaintext);
            }
            catch
            {
                return null;
            }
        }

        /// <summary>
        /// Generuje hash urządzenia na podstawie headers
        /// </summary>
        public static string GenerateDeviceHash(string userAgent, string acceptLanguage)
        {
            var data = JsonSerializer.Serialize(new { userAgent, acceptLanguage });
            var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(data));
            return Convert.ToHexString(bytes).ToLowerInvariant();
        }

        /// <summary>
        /// Generuje hash urządzenia v2 z kontekstu
        /// </summary>
        public static string GenerateDeviceHashV2(Dictionary<string, object?> context)
        {
            var normalized = NormalizeFingerprintContext(context);
            var json = JsonSerializer.Serialize(normalized);
            var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(json));
            return Convert.ToHexString(bytes).ToLowerInvariant();
        }

        private static Dictionary<string, object?> NormalizeFingerprintContext(Dictionary<string, object?> context)
        {
            var normalized = new SortedDictionary<string, object?>();
            foreach (var entry in context)
            {
                if (entry.Value == null)
                    continue;

                if (entry.Value is Dictionary<string, object?> nested)
                {
                    normalized[entry.Key] = NormalizeFingerprintContext(nested);
                }
                else
                {
                    normalized[entry.Key] = entry.Value;
                }
            }

            return new Dictionary<string, object?>(normalized);
        }

        private static string GenerateHotpCode(string secret, long counter, int digits)
        {
            if (counter < 0)
                return new string('0', digits);

            var key = Base32Decode(secret);
            if (key.Length == 0)
                throw new ArgumentException("Nieprawidłowy sekret TOTP", nameof(secret));

            var counterBytes = BitConverter.GetBytes(counter);
            if (BitConverter.IsLittleEndian)
                Array.Reverse(counterBytes);

            using var hmac = new HMACSHA1(key);
            var hash = hmac.ComputeHash(counterBytes);
            var offset = hash[^1] & 0x0f;

            var binary = ((hash[offset] & 0x7f) << 24)
                       | ((hash[offset + 1] & 0xff) << 16)
                       | ((hash[offset + 2] & 0xff) << 8)
                       | (hash[offset + 3] & 0xff);

            var mod = (int)Math.Pow(10, digits);
            var otp = binary % mod;

            return otp.ToString(new string('0', digits));
        }

        private static byte[] Base32Decode(string secret)
        {
            const string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
            var normalized = secret.Replace("=", string.Empty, StringComparison.Ordinal)
                .Replace(" ", string.Empty, StringComparison.Ordinal)
                .ToUpperInvariant();

            var bits = new StringBuilder(normalized.Length * 5);

            foreach (var c in normalized)
            {
                var idx = alphabet.IndexOf(c);
                if (idx < 0)
                    return Array.Empty<byte>();

                bits.Append(Convert.ToString(idx, 2).PadLeft(5, '0'));
            }

            using var stream = new MemoryStream();
            for (var i = 0; i + 8 <= bits.Length; i += 8)
            {
                var chunk = bits.ToString(i, 8);
                stream.WriteByte(Convert.ToByte(chunk, 2));
            }

            return stream.ToArray();
        }
    }

    /// <summary>
    /// Struktura danych w pliku .mijauth
    /// </summary>
    public class AuthPayload
    {
        public string UserId { get; set; } = "";
        public string Token { get; set; } = "";
        public string CreatedAt { get; set; } = "";
        public string? DeviceHash { get; set; }
        public string? DeviceHashV2 { get; set; }
        public int Version { get; set; }
    }

    /// <summary>
    /// Model użytkownika
    /// </summary>
    public class User
    {
        public string Id { get; set; } = "";
        public string Email { get; set; } = "";
        public string PasswordHash { get; set; } = "";
        public string EncryptionKey { get; set; } = "";
        public string AuthToken { get; set; } = "";
        public string TotpSecret { get; set; } = "";
        public string CreatedAt { get; set; } = "";
    }

    /// <summary>
    /// Symulacja bazy danych użytkowników
    /// </summary>
    public class UserDatabase
    {
        private Dictionary<string, User> _users = new();
        private readonly string _storageFile;

        public UserDatabase(string storageFile = "users.json")
        {
            _storageFile = storageFile;
            Load();
        }

        private void Load()
        {
            if (File.Exists(_storageFile))
            {
                var json = File.ReadAllText(_storageFile);
                _users = JsonSerializer.Deserialize<Dictionary<string, User>>(json) ?? new();
            }
        }

        private void Save()
        {
            var json = JsonSerializer.Serialize(_users, new JsonSerializerOptions 
            { 
                WriteIndented = true 
            });
            File.WriteAllText(_storageFile, json);
        }

        public (User User, string AuthFile) CreateUser(string userId, string email, string password)
        {
            var userKey = MijAuthService.GenerateUserKey();
            var (fileContent, token) = MijAuthService.CreateAuthFile(userId, userKey, null, null);
            var totpSecret = MijAuthService.GenerateTotpSecret();

            // Hash hasła (w produkcji użyj BCrypt lub Argon2)
            var passwordHash = Convert.ToHexString(
                SHA256.HashData(Encoding.UTF8.GetBytes(password + "salt_" + userId))
            ).ToLowerInvariant();

            var user = new User
            {
                Id = userId,
                Email = email,
                PasswordHash = passwordHash,
                EncryptionKey = userKey,
                AuthToken = token,
                TotpSecret = totpSecret,
                CreatedAt = DateTime.UtcNow.ToString("o")
            };

            _users[userId] = user;
            Save();

            return (user, fileContent);
        }

        public User? GetUser(string userId)
        {
            return _users.TryGetValue(userId, out var user) ? user : null;
        }

        public User? GetUserByEmail(string email)
        {
            return _users.Values.FirstOrDefault(u => u.Email == email);
        }

        public bool VerifyPassword(User user, string password)
        {
            var hash = Convert.ToHexString(
                SHA256.HashData(Encoding.UTF8.GetBytes(password + "salt_" + user.Id))
            ).ToLowerInvariant();

            return CryptographicOperations.FixedTimeEquals(
                Encoding.UTF8.GetBytes(hash),
                Encoding.UTF8.GetBytes(user.PasswordHash)
            );
        }

        public void UpdateAuthToken(string userId, string newToken)
        {
            if (_users.TryGetValue(userId, out var user))
            {
                user.AuthToken = newToken;
                Save();
            }
        }

        public void DeleteStorage()
        {
            if (File.Exists(_storageFile))
                File.Delete(_storageFile);
        }
    }
}
